### ISSUE TITLE
Style generation for vite virtual modules

### ISSUE THREAD

**Issue by benjaminpreiss** [tailwindlabs/tailwindcss#12696]
**What version of Tailwind CSS are you using?**

3.4.0

**What build tool (or framework if it abstracts the build tool) are you using?**

@sveltejs/kit 1.20.4

**What version of Node.js are you using?**

v20

**What browser are you using?**

Chrome

**What operating system are you using?**

MacOs

**Reproduction URL**

https://replit.com/@benjaminpreiss/tailwindcss-vite-virtual-modules#src/routes/+page.svelte

**Describe your issue**

The REPL shows that styles placed in vite virtual modules are not generated.

My best guess was to write the tailwind config as follows, but that isn't working:

```js
/** @type {import('tailwindcss').Config} */
export default {
  content: ['./src/**/*.{html,js,svelte,ts}', 'virtual:*'],
  theme: {
    extend: {},
  },
  plugins: [],
}
```

**Comment by wongjn** [comment]
[As per the documentation](https://tailwindcss.com/docs/content-configuration):

> The `content` section of your `tailwind.config.js` file is where you configure the paths to all of your HTML templates, JavaScript components, and any other source files that contain Tailwind class names.

Your virtual modules define classes in `vite.config.js`, so you could consider adding this file to your `content`:

```diff
 /** @type {import('tailwindcss').Config} */
 export default {
-  content: ['./src/**/*.{html,js,svelte,ts}', 'virtual:*'],
+  content: ['./src/**/*.{html,js,svelte,ts}', './vite.config.js'],
   theme: {
     extend: {},
```
See it working in this [forked REPL](https://replit.com/@wongjn/httpsgithubcomtailwindlabstailwindcssissues12696).

**Comment by benjaminpreiss** [comment]
Well, this is a workaround, yes. But that doesn't help in my case as I only want tailwind to generate the classes injected by virtual modules.

Your workaround would generate all styles, regardless of what is used. I instead only want to generate what is actually used / imported

**Comment by benjaminpreiss** [comment]
So if I understand correctly, I have several options:

- Delete / add styles to `./vite.config.js`
- Safelist the desired classes in the content config
- Manually `fs.writeFile` to disk from within the vite transformation chain (but where to?)

Are there any other ways to signal to tailwind that I want to safelist certain classes or that I want it to parse certain files (except via the content config)?

**Comment by wongjn** [comment]
> Are there any other ways to signal to tailwind that I want to safelist certain classes […]?

Yes, use the [`safelist` configuration option](https://tailwindcss.com/docs/content-configuration#safelisting-classes).

> Are there any other ways to signal to tailwind […] that I want it to parse certain files (except via the content config)?

Not that I am aware of.

**Comment by benjaminpreiss** [comment]
So that would mean I have to manipulate the tailwind configuration file from within my plugin and add / remove the safelisted classes? Feels a little hacky :(

**Comment by benjaminpreiss** [comment]
Ah, I have another idea... Couldn't I in theory:

- Run the tailwind CLI within vite on the files that I need
- Generate CSS files for each run
- Merge the resulting CSS files with the global CSS

The documentation on the tailwind cli is a little sparse so thankful for any hints

**Comment by thecrypticace** [comment]
**The supported way to do this is to add `vite.config.js` to your content list.** My suggestion is to just add that to your config and not worry about extra potentially unused classes. Tailwind CSS interacts with the filesystem and not with the bundler that you happen to be running it in (there's no way for us to).

Adding a mode to the CLI to spit out a list of classes isn't going to help you because it you'll still have to include classes in your safelist or `vite.config.js` in your content array otherwise there's no way for us to know where to find them. And then you're back in the same boat — including classes that you may potentially not use. IMO, ending up with _some_ extra classes is totally fine and not something you should worry about unless your virtual module is generating tens to hundreds of thousands of rules — but even then you're still going to have to include those classes if used.

If you *really, really* want to do this my suggestion would be a custom vite plugin that runs at the appropriate time, writes a file to disk, that Tailwind CSS can then scan because it's in your content array. Chances are though that this will suffer from a "double build" problem where you have to run a build twice for it to be picked up because it has to first, build all the modules then write the file to disk, and then on the second pass tailwind can now scan the updated file. There's a small chance that the pipeline is set up in such a way that you won't have to but I wouldn't bet on that working forever as it basically relies on implementation details of Vite.

**Comment by benjaminpreiss** [comment]
For reference to the interested reader: https://github.com/tailwindlabs/tailwindcss/discussions/12812

Dynamic safelisting can potentially be made possible by double processing files (once in vite and once in postcss).
